### DISCUSSION TITLE
useFieldArray with MUI AutoComplete Component flickers when new items are appended

### DISCUSSION THREAD

**Discussion by vikasm361** [react-hook-form/react-hook-form#11254] [Q&A]
Hi ,
i have created a Form which uses react-hook-form. The form is dynamic and uses the useFieldArray to add the dropdowns dynamically , the dropdowns are created using the Material UI Autocomplete/TextField combo. There is a button to add the dropdown dynamically on the screen using the append method of useFieldArray. I see that the dropdown flickers as and when i add new dropdowns. This occurs only if i import the FormState.Errors or FormState.isValid in the screen.  Please note the Autocomplete is a controlled component.

This code sample from [codeSandBox ](https://codesandbox.io/p/sandbox/jovial-orla-7nkk9d?layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clpl9rafp0006356h72rj5o3p%2522%252C%2522sizes%2522%253A%255B70%252C30%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clpl9rafo0002356hplyqj8nz%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clpl9rafo0003356hq68luxsx%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clpl9rafo0005356h0fmjg4kf%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B69.3240805409223%252C30.675919459077704%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clpl9rafo0002356hplyqj8nz%2522%253A%257B%2522id%2522%253A%2522clpl9rafo0002356hplyqj8nz%2522%252C%2522tabs%2522%253A%255B%255D%257D%252C%2522clpl9rafo0005356h0fmjg4kf%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clpl9rafo0004356hap4mrknq%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522UNASSIGNED_PORT%2522%252C%2522port%2522%253A0%252C%2522path%2522%253A%2522%2522%257D%255D%252C%2522id%2522%253A%2522clpl9rafo0005356h0fmjg4kf%2522%252C%2522activeTabId%2522%253A%2522clpl9rafo0004356hap4mrknq%2522%257D%252C%2522clpl9rafo0003356hq68luxsx%2522%253A%257B%2522tabs%2522%253A%255B%255D%252C%2522id%2522%253A%2522clpl9rafo0003356hq68luxsx%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A23.125%257D)demonstrates the issue . 
is there a way to avoid this flicker issue? what might be causing it?

**Comment by patorjk** [comment]
I'm pretty sure this is a bug in MUI's Autocomplete control. I was having the exact same issue - I have a form using react-hook-form with useFieldArray and MUI's Autocomplete - and I kept seeing an annoying flicker. After putting in a lot of debug statements, I discovered that MUI passes a value of "" to the Autocomplete's textfield when the control is initially rendered (even though the Autocomplete has a value). Normally this isn't an issue, but if a component above your autocomplete has a "key" prop, it'll completely re-render and you'll see the flicker when this happens. You can see the issue in-depth in this modified version of your codesandbox: (you'll need to open the developer console)

https://codesandbox.io/p/sandbox/objective-cannon-9m3gtz?file=%2Fsrc%2FfieldArray.js

I was able to "fix" it by writing a wrapper component around TextField. This wrapper would check and see if TextField had been touched, and if not it would use a default value that I passed in (rather than ""). Once touched it would start using the value from params.inputProps.value.

Hopefully you were able to find a workaround, I'm mostly writing this for anyone else who comes across this issue. I may file a bug report with MUI later.

**Reply by Shelrothman** [reply]
Could you provide a snippet of the your "fix"? I am curious to see if it would solve my issue as well

**Reply by DanRutledge** [reply]
Thanks for this comment. I was also able to solve my autocomplete flickering/blinking problem by using a wrapper component around the TextField. This was the only helpful solution I found on StackOverflow or Github. Appreciate it @patorjk

**Reply by 1010nishant** [reply]
Could someone provide me with the code i am getting this problem.
link of sandbox is not functioning for me.

**Reply by 1010nishant** [reply]
@DanRutledge and @patorjk, could you please share a code sandbox for this fix? I appreciate your help.

**Reply by DanRutledge** [reply]
@1010nishant It's been a while since I worked on this code, so I'm sorry this is not a completely working code sandbox, but what I did, based on the guidance above is create a wrapper component of the Textfield that forced the value to be what I wanted it to be.  I passed in a 'secretValue' variable and used it to override the value. Hopefully this helps:

const TextWrapper = (props) => {
  const newProps = {...props, inputProps: {...props.inputProps, value: props.secretValue}};
  delete newProps.secretValue;
  return (
    <TextField
    {...newProps}
  />
  )
}

<Autocomplete
   id="myId"
   options={timeList}
   noOptionsText={'Type the format h:mm am/pm'}
   getOptionLabel={(time) => time.label }
   onInputChange={(e, newValue) => {selectTime(newValue);}}
   value={rowStart}
   renderOption={ (props, option) => {
      // jsx does not like the key passed in props
      const key=props.key;
      const propsWithoutKey = {...props};
      delete propsWithoutKey.key;
        return (
          <MenuItem key={key} {...propsWithoutKey}>{option}</MenuItem>
        )                                  
    }}
    renderInput={(params) => {
      // Text wrapper ensures the initial value is what you want it to be
        return (
      <TextWrapper
        secretValue={rowStart}
        {...params}
        variant="standard"
      />
    )}}
  />

**Reply by 1010nishant** [reply]
@DanRutledge  thanks buddy it worked for me!
